consensus/istanbul/validator/weighted.go

```diff
@@ -586,41 +586,88 @@ func (valSet *weightedCouncil) Refresh(hash common.Hash, blockNum uint64, stakin
-	if newStakingInfo.UseGini && newStakingInfo.Gini == reward.DefaultGiniCoefficient && len(newStakingInfo.CouncilNodeAddrs) != 0 {
-		calcGiniCoefficientOfValidators(valSet.validators, newStakingInfo)
+	weightedValidators, stakingAmounts, err := valSet.getStakingAmountsOfValidators(newStakingInfo)
+	if err != nil {
+		return err
 	}
+	totalStaking := calcTotalAmount(newStakingInfo, stakingAmounts)
+	calcWeight(weightedValidators, stakingAmounts, totalStaking)
+
+	valSet.refreshProposers(seed, blockNum)
 
-	// Adjust each validator's staking amount by applying the Gini coefficient if necessary.
-	// Also calculate the total staking amount by summing up the staking amount of each validator.
+	logger.Info("Refresh done.", "blockNum", blockNum, "hash", hash, "valSet.blockNum", valSet.blockNum, "stakingInfo.BlockNum", valSet.stakingInfo.BlockNum)
+	logger.Debug("New proposers calculated", "new proposers", valSet.proposers)
+
+	return nil
+}
+
+// calculate stakingAmounts of validators.
+// if validators have multiple staking contracts, stakingAmounts will be a sum of stakingAmounts with same rewardAddress.
+//  - []*weightedValidator : list of validators which type is converted to weightedValidator
+//  - []float64 : list of stakingAmounts.
+func (valSet *weightedCouncil) getStakingAmountsOfValidators(stakingInfo *reward.StakingInfo) ([]*weightedValidator, []float64, error) {
+	numValidators := len(valSet.validators)
 	weightedValidators := make([]*weightedValidator, numValidators)
 	stakingAmounts := make([]float64, numValidators)
-	totalStaking := float64(0)
+	addedStaking := make([]bool, len(stakingInfo.CouncilNodeAddrs))
+
 	for vIdx, val := range valSet.validators {
 		weightedVal, ok := val.(*weightedValidator)
 		if !ok {
-			return errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
+			return nil, nil, errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
 		}
 		weightedValidators[vIdx] = weightedVal
 
-		i, err := newStakingInfo.GetIndexByNodeId(weightedVal.address)
+		sIdx, err := stakingInfo.GetIndexByNodeId(weightedVal.address)
 		if err == nil {
-			weightedVal.SetRewardAddress(newStakingInfo.CouncilRewardAddrs[i])
-			tempStakingAmount := float64(newStakingInfo.CouncilStakingAmounts[i])
-			if newStakingInfo.UseGini {
-				tempStakingAmount = math.Round(math.Pow(tempStakingAmount, 1.0/(1+newStakingInfo.Gini)))
-			}
-			stakingAmounts[vIdx] = tempStakingAmount
-			totalStaking += tempStakingAmount
+			rewardAddr := stakingInfo.CouncilRewardAddrs[sIdx]
+			weightedVal.SetRewardAddress(rewardAddr)
+			stakingAmounts[vIdx] = float64(stakingInfo.CouncilStakingAmounts[sIdx])
+			addedStaking[sIdx] = true
 		} else {
 			weightedVal.SetRewardAddress(common.Address{})
 		}
 	}
 
-	// Update each validator's weight based on the ratio of its staking amount vs. the total staking amount.
+	for sIdx, isAdded := range addedStaking {
+		if isAdded == false {
```

Comment:
If `if isAdded == false {` is changed like below, we can reduce additional one-depth. Also, you don't need `==` operator for boolean values.

```
if isAdded  {
  continue
}
for vIdx, val := range weightedValidators {
...
}
```

Reply:
reflected